### PR TITLE
feat: [IOPID-315] LV at startup

### PR DESCRIPTION
**Pull request by SabPicariello**
## Short description
This PR, with some mocked clients, enables fast login functionality in startup saga.

## List of changes proposed in this pull request
- ts/components/cie/CieRequestAuthenticationOverlay.tsx: added selector for fast login ff.
- ts/features/fastLogin/backend/index.ts: added functions to perform a fast login.
- ts/features/fastLogin/backend/mockedClients.ts: added mocked clients to perform a fast login.
- ts/features/fastLogin/backend/mockedFunctionsAndTypes.ts: added mocked types to perform a fast login.
- ts/features/fastLogin/saga/tokenRefreshSaga.ts: added saga that perform a complete fast login, from getting Nonce to the actual fast login.
- ts/features/lollipop/hooks/useLollipopLoginSource.tsx: added selector for fast login ff.
- ts/features/lollipop/index.ts: added header for login type (Standard or FL).
- ts/features/lollipop/saga/index.ts: added function to get KeyInfo.
- [ts/features/lollipop/utils/__tests__/retryAndVerifyMismatch.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyMissingSamlRequest.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyMissingThumbprint.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyRedirectNativeError.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifySuccess.test.ts]: Updated tests.
- ts/features/lollipop/utils/fetch.ts: added to lollipopFetch a maxRetries parameter.
- ts/features/lollipop/utils/login.ts: added parameter for Fast Login.
- ts/sagas/__tests__/initializeApplicationSaga.test.ts: updated and added test.
- ts/sagas/index.ts: added saga watch for fast login refresh token.
- ts/sagas/startup.ts: added check on 401 response from getSession.
- ts/screens/authentication/idpAuthSessionHandler.tsx: added fast login selector.
- ts/store/actions/authentication.ts: added action to refresh token for fast login.
- ts/store/reducers/authentication.ts: added reducer to set the new refreshed token.

## How to test
1) Run dev server and configure it for fast login. Set an expiration date for assertionRef too.
2) Run IO app and authenticate.
3) Close the app.
4) Reopen the app after the duration time set for the token in the dev server.
5) It should perform a fast login:
    - getSession : 401
    - getNonce: 200
    - fastLogin: 200
6) Close the app.
7) Reopen the app after the duration time set for the assertionRef in the dev server.
8) It should log you out.

To test the joint retry mechanism for getNonce and fastLogin, with a tool like ProxyMan, try to return error states for those calls

<details><summary>Dev server configuration exmpale</summary>
<p>

```json
{
  "profile": {
    "attrs": {
      "name": "Gian Maria"
    }
  },
  "features": {
    "lollipop": {
      "enabled": true,
      "assertionRefValidityMS": 120000
    },
    "fastLogin": {
      "sessionTTLinMS": 60000
    }
  }
}
```

</p>
</details>

### CODE REVIEW

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  if (errorDescription.status === 403 || errorDescription.status === 401) {
+    // eslint-disable-next-line functional/immutable-data
+    requestState.success = true;
+    return;
+  }
+  // eslint-disable-next-line functional/immutable-data
+  requestState.counter++;
+};
+
+type ErrorDescription = {
```
Comment (shadowsheep1) [inline ts/features/fastLogin/saga/tokenRefreshSaga.ts:130]: Would something like `FastLoginTokenRefreshError` be a more descriptive name than `ErrorDescription`? What do you think about that?

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        lollipopConfig
+      );
+
+      const tokenResponse: SagaCallReturnType<typeof performFastLogin> =
+        yield* call(performFastLogin, fastLoginClient);
+      if (E.isRight(tokenResponse) && tokenResponse.right.status === 200) {
+        // eslint-disable-next-line functional/immutable-data
+        requestState.success = true;
+
+        // TODO: replace the token response type with the correct one.
```
Comment (shadowsheep1) [inline ts/features/fastLogin/saga/tokenRefreshSaga.ts:76]: Here we miss a JIRA ticket link.

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  NonceResponse
+} from "../backend/mockedFunctionsAndTypes";
+
+export function* watchTokenRefreshSaga(): SagaIterator {
+  yield* takeLatest(refreshSessionToken.request, doRefreshTokenSaga);
+}
+
+type RequestStateType = {
+  counter: number;
+  success: boolean;
```
Comment (shadowsheep1) [inline ts/features/fastLogin/saga/tokenRefreshSaga.ts:36]: I wouldn't use a simple `boolean` to  map the "`success`". I would instead use a more explicit "`status`" field like for example `"in-progress" | "succeded" | "failed"` (for example) and I would use a discriminated unions to map the entire `RequestStateType` what do you think?

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      ({ status }) => ({
+        status,
+        description: `response status ${status}`
+      })
+    )
+  );
+  // eslint-disable-next-line functional/immutable-data
+  requestState.error = errorDescription.description;
+
+  if (errorDescription.status === 403 || errorDescription.status === 401) {
```
Comment (shadowsheep1) [inline ts/features/fastLogin/saga/tokenRefreshSaga.ts:121]: Shouldn't we log the citizen out if we get a `403` only?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]
I've suggested some minor changes. Let me know what do you think about them!

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  yield* takeLatest(refreshSessionToken.request, doRefreshTokenSaga);
+}
+
+type RequestStateType = {
+  counter: number;
+  success: boolean;
+  error: string | undefined;
+};
+
+const MAX_RETRYES = 3;
```
Comment (shadowsheep1) [inline ts/features/fastLogin/saga/tokenRefreshSaga.ts:40]: Even if we already don't know how to manage this. Either by a remote config (io-services-metadata) or not, I suggest to put this in our `.env.*` environment files and retrieve this value from configs. Later, if we adopt a remote approach we could use this value as a fallback anyway. What do you think about that?

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts (hunk 3 above)
Comment (shadowsheep1) [reply ts/features/fastLogin/saga/tokenRefreshSaga.ts:36]: Or if you prefer to leave it the way it is, we could just add a comment a track it on with a JIRA ticket.

Comment (SabPicariello) [reply ts/features/fastLogin/saga/tokenRefreshSaga.ts:36]: I think the first solution you proposed, more explicit, is generally the best. I'm going to edit :+1:

Comment (shadowsheep1) [reply ts/features/fastLogin/saga/tokenRefreshSaga.ts:36]: Ops.. the second comment was for the `MAX_RETRIES` part... I put there another comment ;)

File: ts/features/fastLogin/saga/tokenRefreshSaga.ts (hunk 5 above)
Comment (shadowsheep1) [reply ts/features/fastLogin/saga/tokenRefreshSaga.ts:40]: Or if you prefer to leave it the way it is, we could just add a comment and track it on with a JIRA ticket.

**Review by shadowsheep1 — APPROVED** [review]
LGTM

**Pull request merged**